modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1636,16 +1636,16 @@ def apply_full_axis_select_indices(
             self._partition_mgr_cls.apply_func_to_select_indices_along_full_axis(
                 axis,
                 self._partitions,
-                func,
+                self._build_mapreduce_func(axis, func),
                 dict_indices,
                 keep_remaining=keep_remaining,
             )
         )
         # TODO Infer columns and index from `keep_remaining` and `apply_indices`
         if new_index is None:
-            new_index = self.index if axis == 1 else None
+            new_index = self._compute_axis_labels(0, new_partitions)
         if new_columns is None:
-            new_columns = self.columns if axis == 0 else None
+            new_columns = self._compute_axis_labels(1, new_partitions)
```

Comment:
previously, none of the `apply_full_axis_select_indices` calls were hitting these branches (we've been always passing some index and column labels [[1]](https://github.com/modin-project/modin/blob/6550939753c76e896ef2bfd65bb9468d6ad161d7/modin/core/storage_formats/pandas/query_compiler.py#L1570-L1576), [[2]](https://github.com/modin-project/modin/blob/6550939753c76e896ef2bfd65bb9468d6ad161d7/modin/core/storage_formats/pandas/query_compiler.py#L2238-L2245), [[3]](https://github.com/modin-project/modin/blob/6550939753c76e896ef2bfd65bb9468d6ad161d7/modin/core/storage_formats/pandas/query_compiler.py#L2322-L2329)), since this PR these lines are being hit by tests and it appeared that these conditions are incorrect:
1. ModinDataframe can't handle cases when indices/columns are `None`.
2. The method doesn't state that the data shape shouldn't be changed, so we can't preserve old labels.

That's why I replaced this with the explicit labels computation.

Reply:
The original code never works if it hits these branches, so there's actually no stamp to compare with. Making an assumption, that it was supposed to use `apply_indices` instead of `None` here and calling this case "original", the performance difference is the following:
```
========== implementation='original' ==========
narrow frame (10000000, 10): 1.495953186787665s.
wide frame (5000, 5000): 0.1285325076431036s.

========== implementation='#3944' ==========
narrow frame (10000000, 10): 1.3315681890584528s.
wide frame (5000, 5000): 0.18555662231519818s.
```

<details><summary>Script that measured it</summary>

<details><summary>Hacky injection to make comparison</summary>

```diff
diff --git a/modin/core/dataframe/pandas/dataframe/dataframe.py b/modin/core/dataframe/pandas/dataframe/dataframe.py
index aea8f532..259ad11a 100644
--- a/modin/core/dataframe/pandas/dataframe/dataframe.py
+++ b/modin/core/dataframe/pandas/dataframe/dataframe.py
@@ -1596,6 +1596,7 @@ class PandasDataframe(object):
         new_index=None,
         new_columns=None,
         keep_remaining=False,
+        __implementation__=None,
     ):
         """
         Apply a function across an entire axis for a subset of the data.
@@ -1643,9 +1644,19 @@ class PandasDataframe(object):
         )
         # TODO Infer columns and index from `keep_remaining` and `apply_indices`
         if new_index is None:
-            new_index = self._compute_axis_labels(0, new_partitions)
+            if __implementation__ == "original":
+                new_index = self.index if axis == 0 else apply_indices
+            elif __implementation__ == "#3944":
+                new_index = self._compute_axis_labels(0, new_partitions)
+            else:
+                assert False
         if new_columns is None:
-            new_columns = self._compute_axis_labels(1, new_partitions)
+            if __implementation__ == "original":
+                new_columns = self.columns if axis == 1 else apply_indices
+            elif __implementation__ == "#3944":
+                new_columns = self._compute_axis_labels(1, new_partitions)
+            else:
+                assert False
         return self.__constructor__(new_partitions, new_index, new_columns, None, None)

     @lazy_metadata_decorator(apply_axis="both")
```

</details>

```python
import modin.pandas as pd
import numpy as np
import timeit

shapes = {
    "narrow frame": (10_000_000, 10),
    "wide frame": (5000, 5000),
}


def to_measure(dataframe, impl):
    return dataframe.apply_full_axis_select_indices(
        axis=0,
        func=lambda df, internal_indices: df.iloc[:, internal_indices] + 10,
        apply_indices=["col1", "col2", f"col{NCOLS - 1}"],
        __implementation__=impl,
    )


for implementation in ("original", "#3944"):
    print(f"========== {implementation=} ==========")
    for key, shape in shapes.items():
        NROWS, NCOLS = shape
        dataframe = pd.DataFrame(
            {f"col{i}": np.arange(NROWS) for i in range(NCOLS)}
        )._query_compiler._modin_frame
        print(
            f"{key} ({NROWS}, {NCOLS}): {timeit.timeit(lambda: to_measure(dataframe, implementation), number=10) / 10}"
        )


```

</details>

Surprisingly, the numbers are literally the same (looks really strange to me, but I double-checked the results and they're right).